src/emulator/storage/files.ts

```diff
@@ -334,55 +283,47 @@ export class StorageLayer {
   }
 
   /**
-   * Stores the uploaded file with generated metadata and triggers Object Finalize Cloud Functions.
-   * @param upload The upload to finalize.
-   * @returns The stored file.
+   * Last step in uploading a file. Validates the request and persists the staging
+   * object to its permanent location on disk.
+   * TODO(tonyjhuang): Inject a Rules evaluator into StorageLayer to avoid needing skipAuth param
+   * @throws {ForbiddenError} if the request fails security rules auth.
    */
-  public finalizeUpload(upload: ResumableUpload): StoredFile {
-    upload.status = UploadStatus.FINISHED;
+  public async handleUploadObject(upload: Upload, skipAuth = false): Promise<StoredFileMetadata> {
+    if (upload.status !== UploadStatus.FINISHED) {
+      throw new Error(`Unexpected upload status encountered: ${upload.status}.`);
+    }
 
-    const metadata = this.createMetadata(upload);
     const filePath = this.path(upload.bucketId, upload.objectId);
-    const file = new StoredFile(metadata, filePath);
-
-    this._files.set(filePath, file);
-
-    this._persistence.deleteFile(filePath, /* failSilently = */ true);
-    this._persistence.renameFile(upload.fileLocation, filePath);
-
-    this._cloudFunctions.dispatch("finalize", new CloudStorageObjectMetadata(file.metadata));
-
-    return file;
-  }
-
-  public oneShotUpload(
-    bucket: string,
-    object: string,
-    contentType: string,
-    incomingMetadata: IncomingMetadata,
-    bytes: Buffer
-  ) {
-    const filePath = this.path(bucket, object);
-
-    this._persistence.deleteFile(filePath, true);
-
-    this._persistence.appendBytes(filePath, bytes);
-    const md = new StoredFileMetadata(
+    const metadata = new StoredFileMetadata(
       {
-        name: object,
-        bucket: bucket,
-        contentType: incomingMetadata.contentType || "application/octet-stream",
-        contentEncoding: incomingMetadata.contentEncoding,
-        customMetadata: incomingMetadata.metadata,
+        name: upload.objectId,
+        bucket: upload.bucketId,
+        contentType: upload.metadata.contentType || "application/octet-stream",
+        contentEncoding: upload.metadata.contentEncoding,
+        customMetadata: upload.metadata.metadata,
```

Comment:
Nit, slight rename from `metadata` -> `incomingMetadata` as it was originally or even `uploadMetadata`? ie. `upload.incomingMetadata.metadata`

Given we have a couple different metadatas being used from the `metadata.ts` file, might be good for readability.

Reply:
Pushing back a little, given this particular metadata field belongs to the `Upload` type, IMO it's relatively clear that `Upload.metadata` is representative of incoming metadata. Do you think renaming the metadata var on line 297 to `storedMetadata` would be sufficient?